File: src/scripts/lint_featured_skills.py
#!/usr/bin/env python3
"""CI guard for docs/featured-skills.md entry validity.

Every artefact referenced from the three Featured tables (Founders &
Consultants, Content Creators, Engineering Leads) MUST resolve in
`dist/discovery/discovery-manifest.json`. Stale entries (renamed or
removed skills / commands) fail the build.

Detection:

  - Scan `docs/featured-skills.md` for inline links of shape
    `[`<token>`](../.agent-src/{skills|commands}/<path>.md)` inside the
    Featured tables. Strip the `/` prefix on commands and the leading
    slash on skill names.
  - Cross-check each token against the manifest's `artefacts` array
    (`category` in {`skill`, `command`} + `name` match, namespaced
    commands like `video/from-script` → `video:from-script`).
  - Verify `--pack <slug>` install hints reference packs that exist in
    `manifest.packs[].id`.

Exit codes:
  0 — every entry resolves; install-pack hints are valid.
  1 — at least one stale entry or unknown pack.

Invocation:
  python3 scripts/lint_featured_skills.py
  python3 scripts/lint_featured_skills.py --quiet
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

DOC = Path("docs/featured-skills.md")
MANIFEST = Path("dist/discovery/discovery-manifest.json")

# Matches `[`token`](../.agent-src/skills/<slug>/SKILL.md)` or
# `[`/token`](../.agent-src/commands/<path>.md)`. Captures (category, slug-path).
LINK_RE = re.compile(
    r"\[`/?[^`]+`\]\(\.\./\.agent-src/(skills|commands)/([^)]+?)\.md\)"
)
PACK_HINT_RE = re.compile(r"--pack\s+([a-z][a-z0-9-]*)")


def load_manifest() -> dict:
    if not MANIFEST.exists():
        print(f"error: manifest not found at {MANIFEST}", file=sys.stderr)
        sys.exit(1)
    return json.loads(MANIFEST.read_text(encoding="utf-8"))


def manifest_names(manifest: dict) -> tuple[set[str], set[str], set[str]]:
    """Return (skill-names, command-names, pack-ids)."""
    skills: set[str] = set()
    commands: set[str] = set()
    for art in manifest.get("artefacts", []):
        cat = art.get("category")
        name = art.get("name")
        if not name:
            continue
        if cat == "skill":
            skills.add(name)
        elif cat == "command":
            commands.add(name)
            # Deprecation aliases are permanent stubs (ADR-057 § 8a) and
            # still resolve — docs may feature a command by its alias.
            for alias in art.get("replaces") or []:
                if isinstance(alias, str) and alias:
                    commands.add(alias)
    packs = {p.get("id") for p in manifest.get("packs", []) if p.get("id")}
    return skills, commands, packs


def slug_from_path(category: str, raw: str) -> str:
    """Convert a doc-link path to the manifest `name` form.

    skills/<slug>/SKILL  → <slug>
    commands/<group>/<leaf>  → <group>:<leaf>
    commands/<leaf>     → <leaf>
    """
    if category == "skills":
        # raw looks like "<slug>/SKILL"; strip trailing /SKILL if present.
        return raw.split("/", 1)[0]
    # commands
    parts = raw.split("/")
    return ":".join(parts) if len(parts) > 1 else parts[0]


def main() -> int:
    quiet = "--quiet" in sys.argv
    if not DOC.exists():
        print(f"error: {DOC} not found", file=sys.stderr)
        return 1

    manifest = load_manifest()
    skills, commands, packs = manifest_names(manifest)
    body = DOC.read_text(encoding="utf-8")

    missing: list[str] = []
    seen: set[tuple[str, str]] = set()
    for cat, raw in LINK_RE.findall(body):
        slug = slug_from_path(cat, raw)
        key = (cat, slug)
        if key in seen:
            continue
        seen.add(key)
        pool = skills if cat == "skills" else commands
        if slug not in pool:
            missing.append(f"  - {cat}/{slug} (linked path: ../.agent-src/{cat}/{raw}.md)")

    unknown_packs: list[str] = []
    for pack in PACK_HINT_RE.findall(body):
        if pack not in packs:
            unknown_packs.append(f"  - --pack {pack}")

    if missing or unknown_packs:
        print(f"FAIL  {DOC}: stale Featured Skills entries detected.")
        if missing:
            print("\nMissing artefacts (not in discovery-manifest.json):")
            for line in missing:
                print(line)
        if unknown_packs:
            print("\nUnknown pack ids referenced in install hints:")
            for line in unknown_packs:
                print(line)
        print(
            "\nFix: either restore the artefact, update the doc entry to a "
            "current name, or substitute with the nearest existing artefact."
        )
        return 1

    if not quiet:
        print(
            f"OK    {DOC}: {len(seen)} artefact entries + "
            f"{len(set(PACK_HINT_RE.findall(body)))} pack hints validated."
        )
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/scripts/lint_featured_skills.ts
#!/usr/bin/env tsx
/**
 * CI guard for docs/featured-skills.md entry validity.
 *
 * Ported from the retired Python `src/scripts/lint_featured_skills.py` (ADR-088,
 * Phase 4 / Wave 4b). The CLI contract is pinned — `--quiet`
 * detected by argv membership, exit codes (0 / 1), stdout/stderr split,
 * byte-identical messages, same regex scan + ordering. No behaviour
 * changes — historical quirks preserved (consumers pin the exact behaviour).
 *
 * NOTE: like the retired Python implementation, the doc/manifest paths are CWD-relative
 * (`Path("docs/featured-skills.md")`), not `__file__`-anchored. The
 * dispatcher invokes from the repo root, so this matches.
 *
 * Exit codes:
 *   0 — every entry resolves; install-pack hints are valid.
 *   1 — at least one stale entry or unknown pack.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

import { assertScanned, DeadScopeError } from './_lib/scan_scope.js';

const _HERE = fileURLToPath(import.meta.url);

// CWD-relative, exactly as the Python module's Path(...) literals.
const DOC = 'docs/featured-skills.md';
const MANIFEST = 'dist/discovery/discovery-manifest.json';

// Matches `[`token`](../dist/agent-src/skills/<slug>/SKILL.md)` or
// `[`/token`](../dist/agent-src/commands/<path>.md)`. Captures (category, slug-path).
const LINK_RE = /\[`\/?[^`]+`\]\(\.\.\/dist\/agent-src\/(skills|commands)\/([^)]+?)\.md\)/g;
const PACK_HINT_RE = /--pack\s+([a-z][a-z0-9-]*)/g;

interface Manifest {
    artefacts?: Array<Record<string, unknown>>;
    packs?: Array<Record<string, unknown>>;
}

function _exists(p: string): boolean {
    try {
        fs.statSync(p);
        return true;
    } catch {
        return false;
    }
}

function load_manifest(): Manifest {
    if (!_exists(MANIFEST)) {
        process.stderr.write(`error: manifest not found at ${MANIFEST}\n`);
        process.exit(1);
    }
    return JSON.parse(fs.readFileSync(MANIFEST, 'utf-8')) as Manifest;
}

/** Return [skill-names, command-names, pack-ids]. */
function manifest_names(manifest: Manifest): [Set<string>, Set<string>, Set<string>] {
    const skills = new Set<string>();
    const commands = new Set<string>();
    for (const art of manifest.artefacts ?? []) {
        const cat = art['category'];
        const name = art['name'];
        if (!name || typeof name !== 'string') {
            continue;
        }
        if (cat === 'skill') {
            skills.add(name);
        } else if (cat === 'command') {
            commands.add(name);
            // Deprecation aliases are permanent stubs (ADR-057 § 8a).
            const replaces = art['replaces'];
            if (Array.isArray(replaces)) {
                for (const alias of replaces) {
                    if (typeof alias === 'string' && alias) {
                        commands.add(alias);
                    }
                }
            }
        }
    }
    const packs = new Set<string>();
    for (const p of manifest.packs ?? []) {
        const id = p['id'];
        if (id && typeof id === 'string') {
            packs.add(id);
        }
    }
    return [skills, commands, packs];
}

function slug_from_path(category: string, raw: string): string {
    if (category === 'skills') {
        return raw.split('/', 1)[0]!;
    }
    // Nested command files (commands/video/scene.md) carry dash-joined
    // frontmatter names (video-scene) — the form the discovery manifest
    // indexes. The colon form only exists as the user-facing slash label.
    const parts = raw.split('/');
    return parts.length > 1 ? parts.join('-') : parts[0]!;
}

/** All [category, slug-path] matches in document order (mirrors re.findall). */
function _findLinks(body: string): Array<[string, string]> {
    const out: Array<[string, string]> = [];
    LINK_RE.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = LINK_RE.exec(body)) !== null) {
        out.push([m[1]!, m[2]!]);
    }
    return out;
}

function _findPackHints(body: string): string[] {
    const out: string[] = [];
    PACK_HINT_RE.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = PACK_HINT_RE.exec(body)) !== null) {
        out.push(m[1]!);
    }
    return out;
}

function main(argv: readonly string[]): number {
    const quiet = argv.includes('--quiet');
    if (!_exists(DOC)) {
        process.stderr.write(`error: ${DOC} not found\n`);
        return 1;
    }

    const manifest = load_manifest();
    const [skills, commands, packs] = manifest_names(manifest);
    const body = fs.readFileSync(DOC, 'utf-8');

    // The doc existing is not the doc being readable by this gate: LINK_RE is
    // pinned to the `../dist/agent-src/{skills,commands}/…` link shape, so a
    // restructured or emptied Featured Skills page yields zero entries and the
    // OK line reports "0 artefact entries validated". Assert the raw match list,
    // before the dedupe the checks below run on.
    const links = _findLinks(body);
    try {
        assertScanned({
            gate: 'lint_featured_skills',
            scanned: links.length,
            units: 'featured artefact link(s)',
            roots: [DOC],
        });
    } catch (exc) {
        if (exc instanceof DeadScopeError) {
            // 1 is the gate's only failure code — the same one a missing DOC or
            // manifest already returns.
            process.stderr.write(`❌  ${exc.message}\n`);
            return 1;
        }
        throw exc;
    }

    const missing: string[] = [];
    const seen = new Set<string>();
    for (const [cat, raw] of links) {
        const slug = slug_from_path(cat, raw);
        const key = `${cat}\0${slug}`;
        if (seen.has(key)) {
            continue;
        }
        seen.add(key);
        const pool = cat === 'skills' ? skills : commands;
        if (!pool.has(slug)) {
            missing.push(`  - ${cat}/${slug} (linked path: ../dist/agent-src/${cat}/${raw}.md)`);
        }
    }

    const unknown_packs: string[] = [];
    for (const pack of _findPackHints(body)) {
        if (!packs.has(pack)) {
            unknown_packs.push(`  - --pack ${pack}`);
        }
    }

    if (missing.length || unknown_packs.length) {
        process.stdout.write(`FAIL  ${DOC}: stale Featured Skills entries detected.\n`);
        if (missing.length) {
            process.stdout.write('\nMissing artefacts (not in discovery-manifest.json):\n');
            for (const line of missing) {
                process.stdout.write(line + '\n');
            }
        }
        if (unknown_packs.length) {
            process.stdout.write('\nUnknown pack ids referenced in install hints:\n');
            for (const line of unknown_packs) {
                process.stdout.write(line + '\n');
            }
        }
        process.stdout.write(
            '\nFix: either restore the artefact, update the doc entry to a ' +
                'current name, or substitute with the nearest existing artefact.\n',
        );
        return 1;
    }

    if (!quiet) {
        const packHintSet = new Set(_findPackHints(body));
        process.stdout.write(
            `OK    ${DOC}: ${seen.size} artefact entries + ` +
                `${packHintSet.size} pack hints validated.\n`,
        );
    }
    return 0;
}

function _isCliEntry(): boolean {
    if (process.argv[1] === undefined) {
        return false;
    }
    const argvUrl = pathToFileURL(path.resolve(process.argv[1])).href;
    if (import.meta.url === argvUrl) {
        return true;
    }
    // A symlinked invocation (e.g. via an installed `.augment/` projection,
    // or macOS /var → /private/var temp dirs) makes the raw URLs differ:
    // import.meta.url is the resolved real path while argv[1] keeps the
    // symlink path. Compare realpaths so the entry guard still fires
    // (without this the CLI silently no-ops when run through a symlink).
    try {
        const here = fs.realpathSync(fileURLToPath(import.meta.url));
        const argv = fs.realpathSync(path.resolve(process.argv[1]));
        return here === argv;
    } catch {
        return false;
    }
}

if (_isCliEntry() || process.argv[1] === _HERE) {
    process.exit(main(process.argv.slice(2)));
}

export {
    type Manifest,
    DOC,
    MANIFEST,
    LINK_RE,
    PACK_HINT_RE,
    manifest_names,
    slug_from_path,
    main,
};
